[Tania Thng] iP

taniathng opened the pull request:
# Hannah

> “Your mind is for having ideas, not holding them.” – David Allen 

Hannah has a ton of features for you to try!! 

- **Text-based**
- **Easy to learn**
- **FAST & SUPER FAST** to use

All you need to do is:
1. Download it from [here](https://github.com/taniathng/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you! 😉

### Features:
- [x] Managing ToDos
- [x] Managing Deadlines
- [ ] Managing Events
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

You can also edit your code using `inline code` like this: `TaskList.addTask()`.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

taniathng committed:
Level-0: Rename, Greet, Exit

taniathng committed:
Level-1: Add echo functionality

taniathng committed:
Level-2: Add, List

taniathng committed:
Level-3: Mark as Done

taniathng committed:
Level-4: ToDos, Events, Deadlines

taniathng committed:
Leve-5 Handle errors

taniathng committed:
Level-6 Delete

taniathng committed:
Level 7: Save functionality

taniathng committed:
Level 7: Save functionality

taniathng committed:
Merge branch 'branch-Level-'

taniathng committed:
Level 8: Dates and Times

taniathng committed:
Merge branch-Level-8: Dates and Times

taniathng committed:
Extracting UI

taniathng committed:
Extract TaskList

taniathng committed:
Merge branch 'TaskList'

taniathng committed:
Extract Command and Parser Class

taniathng committed:
Merge branch 'CommandClasses'

taniathng committed:
Organize classes into packages

taniathng committed:
Merge branch 'organize-packages'

taniathng committed:
Merge branch 'add-gradle-support'

taniathng committed:
Add Gradle Support

daronoh commented on src/main/java/hannah/Hannah.java:
```diff
@@ -0,0 +1,137 @@
+    private Storage storage;
+    private TaskList list;
+    private Parser parser;
+    private static final String FILE_PATH = "src/data/Hannah.txt";
```
I like the use of the all caps here to name the constant

daronoh commented on src/main/java/hannah/Hannah.java:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+
+    public void run() {
```
Would it be more readable if the lines of code were separated by their functionality?

daronoh commented on src/main/java/hannah/Parser.java:
```diff
@@ -0,0 +1,32 @@
+        String[] parts = userInput.split(" ", 2);
+        String commandType = parts[0];
+
+        switch (commandType) {
+            case "bye":
+                return new Command("bye");
+            case "list":
+                return new Command("list");
+            case "mark":
+                return new Command("mark", parts.length > 1 ? parts[1] : null);
+            case "unmark":
+                return new Command("unmark", parts.length > 1 ? parts[1] : null);
+            case "delete":
+                return new Command("delete", parts.length > 1 ? parts[1] : null);
+            case "todo":
+                return new Command("todo", parts.length > 1 ? parts[1] : null);
+            case "deadline":
+                return new Command("deadline", parts.length > 1 ? parts[1] : null);
+            case "event":
+                return new Command("event", parts.length > 1 ? parts[1] : null);
+            default:
+                return new Command("unknown");
```
Is the indentation level of the case here following the styling standards of the course?

daronoh commented on src/main/java/hannah/Storage.java:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+//    public void updateDoneStatus(hannah.task.Task task) throws IOException {
+//        File file = new File(filePath);
+//        File tempFile = new File("tempFile.txt");
+//
+//        try (BufferedReader reader = new BufferedReader(new FileReader(file));
+//             BufferedWriter writer = new BufferedWriter(new FileWriter(tempFile))) {
+//
+//            String currentLine;
+//            while ((currentLine = reader.readLine()) != null) {
+//                String[] taskInfo = currentLine.split(" \\| ");
+//                if (taskInfo[2].equals(task.getName())) { // Match by task name
+//                    writer.write(task.saveTaskStatus());
+//                } else {
+//                    writer.write(currentLine);
+//                }
+//                writer.newLine();
+//            }
+//        }
+//    }
```
Should commented out code be removed before submitting a PR?

daronoh commented on src/main/java/hannah/task/Deadlines.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Deadlines extends Task {
```
Should deadlines be plural here?

daronoh commented on src/main/java/hannah/task/Events.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+public class Events extends Task {
```
Should events be plural here?

daronoh reviewed (commented):
LGTM. Nice Work!

taniathng committed:
Add JUnit tests for Storage and TaskList classes

taniathng committed:
Merge branch 'add-junit-tests'

taniathng committed:
Added JavaDoc comments to relevant classes and methods

taniathng committed:
Implement feature X with coding standard

taniathng committed:
Implement find feature for searching

taniathng committed:
Merge branch 'branch-A-CodingStandard'

taniathng committed:
Fix merge conflict

taniathng committed:
Implement find feature to search tasks by keyword

gohsl99 commented on src/main/java/hannah/Hannah.java:
```diff
@@ -0,0 +1,164 @@
+     * @param date The date string to validate.
+     * @return true if the date is valid, false otherwise.
+     */
+    private boolean validateDate(String date) {
```
Perhaps method name like "isValidDate" will be more understandable.

gohsl99 commented on src/main/java/hannah/Hannah.java:
```diff
@@ -0,0 +1,164 @@
+                System.out.println("Commands: todo, deadline, event, delete, mark, unmark");
+            }
+            System.out.println("____________________________________________________________");
+        }
```
Good job on naming variables clearly.
Additionally, not a coding standard issue, but you may consider to use switch to replace if else condition, and .equals() is more accurate for String comparison.

gohsl99 commented on src/main/java/hannah/Parser.java:
```diff
@@ -0,0 +1,43 @@
+        String[] parts = userInput.split(" ", 2);
+        String commandType = parts[0];
+
+        switch (commandType) {
+            case "bye":
+                return new Command("bye");
+            case "list":
+                return new Command("list");
+            case "mark":
+                return new Command("mark", parts.length > 1 ? parts[1] : null);
+            case "unmark":
+                return new Command("unmark", parts.length > 1 ? parts[1] : null);
+            case "delete":
+                return new Command("delete", parts.length > 1 ? parts[1] : null);
+            case "todo":
+                return new Command("todo", parts.length > 1 ? parts[1] : null);
+            case "deadline":
+                return new Command("deadline", parts.length > 1 ? parts[1] : null);
+            case "event":
+                return new Command("event", parts.length > 1 ? parts[1] : null);
+            case "find":
+                return new Command("find", parts.length > 1 ? parts[1] : null);
+            default:
+                return new Command("unknown");
+        }
```
Perhaps case and default clauses should have no indentation (if follow the coding standard strictly).

gohsl99 commented on src/main/java/hannah/Storage.java:
```diff
@@ -0,0 +1,132 @@
+import hannah.task.TaskList;
+import hannah.task.Task;
+
+import java.io.*;
```
Perhaps just import the specific class will do, instead of *.

gohsl99 commented on src/main/java/hannah/Storage.java:
```diff
@@ -0,0 +1,132 @@
+
+import hannah.task.Deadline;
+import hannah.task.Event;
+import hannah.task.ToDos;
+import hannah.task.TaskList;
+import hannah.task.Task;
```
Perhaps you can rearrange the import statement alphabetically so it will be more organized.

gohsl99 commented on src/main/java/hannah/Hannah.java:
```diff
@@ -0,0 +1,164 @@
+import java.time.format.DateTimeParseException;
+import java.util.List;
+import java.util.Scanner;
+import java.io.*;
```
Perhaps just import the specific class will do, instead of *.

gohsl99 commented on src/main/java/hannah/Storage.java:
```diff
@@ -0,0 +1,132 @@
+        String type = taskInfo[0];
+        boolean isDone = taskInfo[1].equals("1");
+
+        switch (type) {
+            case "T":
+                ToDos todoTask = new ToDos(taskInfo[2]);
+                if (isDone) {
+                todoTask.setDone();
+                }
+                return todoTask;
+            case "D":
+                Deadline deadlineTask = new Deadline(taskInfo[2]);
+                if (isDone) {
+                    deadlineTask.setDone();
+                }
+                deadlineTask.setDeadline(taskInfo[3].substring(3));
+                return deadlineTask;
+            case "E":
+                Event eventTask = new Event(taskInfo[2]);
+                if (isDone) {
+                    eventTask.setDone();
+                }
+                String duration = taskInfo[3];
+                int fromIndex = duration.indexOf("from ");
+                int toIndex = duration.indexOf("to ");
+                eventTask.setDuration(duration.substring(fromIndex + 5, toIndex-1), duration.substring(toIndex + 3));
+                return eventTask;
+            default:
+                return null;  // Handle any unexpected task types
+        }
```
Perhaps case and default clauses should have no indentation (if follow the coding standard strictly).

gohsl99 commented on src/main/java/hannah/Ui.java:
```diff
@@ -0,0 +1,144 @@
+package hannah;
+
+import java.util.List;
+import java.util.Scanner;
+import hannah.task.Task;
+import hannah.task.TaskList;
```
Perhaps you can separate system and custom import statements, will look more organized.

gohsl99 commented on src/main/java/hannah/Ui.java:
```diff
@@ -0,0 +1,144 @@
+
+    /**
+     * Reads the user's command from the input.
+     *
+     * @return The user's input as a String.
+     */
+    public String readCommand() {
+        return scanner.nextLine();
+    }
+
+    /**
+     * Extracts the task number from the user's input.
+     *
+     * @param userInput The user's input string.
+     * @return The task number extracted from the input.
+     */
+    public int getTaskNumber(String userInput){
+        String[] parts = userInput.split(" ");
+        int taskNumber = 0;
+        if (parts.length > 1) {
+            taskNumber = Integer.parseInt(parts[1]);
+        }
+        return taskNumber;
+    }
+
+    public String getKeyword(String userInput){
+        String[] parts = userInput.split(" ");
+        return parts[1];
+    }
+
+    public void showFindResults(TaskList tasks, String keyword) {
+        System.out.println("Here are the matching tasks in your list:");
+        List<Task> matchingTasks = tasks.findTasks(keyword);
+        for (int i = 0; i < tasks.size(); i++) {
+            System.out.println((i + 1) + "." + matchingTasks.get(i));
+        }
+        System.out.println("____________________________________________________________");
+    }
+
+}
```
Good job, rest of the code looks good!

gohsl99 commented on src/main/java/hannah/task/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+import java.time.format.DateTimeParseException;
+
+public class Deadline extends Task {
+    String name;
+    LocalDate deadline;
+    public Deadline(String name){
+        super(name);
+    }
+    @Override
+    public String toString(){
+        return("[D]" + super.toString() + " by " +  this.deadline);
+    }
```
Not a coding standard issue, perhaps you will want to separate line 9 to 10 and line 12 to 13.

gohsl99 commented on src/main/java/hannah/task/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+        return("[D]" + super.toString() + " by " +  this.deadline);
+    }
+
+    @Override
+    public void setDeadline(String deadline) {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+        try {
+            this.deadline = LocalDate.parse(deadline, formatter);
+        } catch (DateTimeParseException e) {
+            System.out.println("Invalid date format. Please use yyyy-MM-dd.");
+        }
+    }
+    public LocalDate getDeadline(){
+        return this.deadline;
+    }
+    @Override
+    public String saveTaskFormat(){
+        String line = "D | " + (isDone ? "1" : "0") + " | " + super.name + " | by " + this.deadline;
+        return line + System.lineSeparator();
+    }
```
Not a coding standard issue, perhaps you will want to separate line 26 to 27 and line 29 to 30.

gohsl99 commented on src/main/java/hannah/task/Event.java:
```diff
@@ -0,0 +1,37 @@
+import java.time.format.DateTimeParseException;
+
+public class Event extends Task {
+    String name;
+    LocalDate startDuration;
+    LocalDate endDuration;
+    public Event(String name){
+        super(name);
+
+    }
+    @Override
+    public String toString(){
+        return("[E]" + super.toString() + " from " +  this.startDuration + " to " + this.endDuration);
+    }
```
Not a coding standard issue, perhaps you will want to separate line 10 to 11 and line 14 to 15.

gohsl99 commented on src/main/java/hannah/task/Event.java:
```diff
@@ -0,0 +1,37 @@
+    public String saveTaskFormat(){
+        String line = "E | " + (isDone ? "1" : "0") + " | " + super.name + " | from " + this.startDuration + " to " + this.endDuration;
+        return line + System.lineSeparator();
+
```
Not a coding standard issue, perhaps you will want to remove empty line at line 35.

gohsl99 commented on src/main/java/hannah/task/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    /**
+     * Marks a task as done.
+     *
+     * @param task The task to be marked as done.
+     */
+    public void markTask(Task task){
+        if (!task.isDone) {
+            task.setDone();
+        }
+    }
+
+    /**
+     * Unmarks a task (marks it as not done).
+     *
+     * @param task The task to be unmarked.
+     */
+    public void unmarkTask(Task task){
+        if (task.isDone) {
+            task.setDone();
+        }
+    }
+
+    /**
+     * Finds and returns a list of tasks that contain the given keyword.
+     *
+     * @param keyword The keyword to search for.
+     * @return A list of matching tasks.
+     */
+    public List<Task> findTasks(String keyword) {
+        List<Task> matchingTasks = new ArrayList<>();
+        for (Task task : tasks) {
+            if (task.getName().contains(keyword)) {
+                matchingTasks.add(task);
+            }
+        }
+        return matchingTasks;
+    }
+}
```
Good job for JavaDoc and naming convention.

gohsl99 commented on src/test/java/hannah/StorageTest.java:
```diff
@@ -0,0 +1,49 @@
+package hannah;
+
+import hannah.task.Deadline;
+import hannah.task.Task;
+import hannah.task.TaskList;
+import hannah.task.ToDos;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
```
Perhaps you can separate system and custom import statements, will look more organized.

gohsl99 commented on src/test/java/hannah/StorageTest.java:
```diff
@@ -0,0 +1,49 @@
+import java.io.IOException;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps just import the specific class will do, instead of *.

gohsl99 commented on src/test/java/hannah/TaskListTest.java:
```diff
@@ -0,0 +1,42 @@
+import hannah.task.ToDos;
+import hannah.task.Deadline;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Perhaps just import the specific class will do, instead of *.

gohsl99 reviewed (commented):
Overall LGTM, coding standard is follow strictly in most of the code, just some minor improvements can be made. Keep up the good work!

taniathng committed:
Fix Checkstyle issues and improve code formatting

taniathng committed:
add GUI

taniathng committed:
Integrate GUI with Hannah chatbot logic

taniathng committed:
Generated JAR file with dependencies

taniathng committed:
Added assertions for A-Assertions increment

taniathng committed:
Improve code quality for A-CodeQuality increment

taniathng committed:
Merge branch 'master' into A-CodeQuality

taniathng committed:
Merge pull request #1 from taniathng/A-CodeQuality

A code quality

taniathng committed:
change magic string

taniathng committed:
Add help message

taniathng committed:
A-BetterGui: Improve GUI design and add background change images

taniathng committed:
A-BetterGui: Finalize GUI improvements and clean up code

Removed unnecessary comments and polished the user interface for a cleaner appearance. Ensured the background image is properly integrated and improved the overall look and feel of the application.

taniathng committed:
Add JavaDocs

taniathng committed:
Change DialogBox colour

taniathng committed:
Change DialogBox colour

taniathng committed:
Remove lines surrounding messages

taniathng committed:
Delete images that are not used

taniathng committed:
Show welcome message

taniathng committed:
Add Ui.png to docs

taniathng committed:
Add Ui.png to docs and change README

taniathng committed:
Change README

taniathng committed:
Edit Storage and gradle to work with jar file

taniathng committed:
Add exit after Bye and update dependencies
